File: Tutorials/Java/Download2GoAndroidX1_6/src/main/java/com/penthera/download2goX1_6/ui/MainActivity.java
/*
    Created by Penthera on 26/03/21.
    Copyright © 2021 penthera. All rights reserved.

    This source file contains a very basic example showing how to use the Penthera Download2Go SDK.
*/
package com.penthera.download2goX1_6.ui;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import com.penthera.download2goX1_6.AssetsAdapter;
import com.penthera.download2goX1_6.viewmodel.AssetListViewModel;
import com.penthera.download2goX1_6.R;
import com.penthera.common.Common.AuthenticationStatus;
import com.penthera.virtuososdk.androidxsupport.VirtuosoLiveDataFactory;
import com.penthera.virtuososdk.client.ISegmentedAsset;
import com.penthera.virtuososdk.client.ISegmentedAssetFromParserObserver;
import com.penthera.virtuososdk.client.Virtuoso;
import com.penthera.virtuososdk.client.builders.HLSAssetBuilder;
import com.penthera.virtuososdk.client.builders.MPDAssetBuilder;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.UUID;

import static android.content.Intent.ACTION_VIEW;

/**
 * A simple activity containing buttons to create three assets and a RecyclerView to show all stored assets.
 * The RecyclerView is populated using a basic ViewModel, which uses LiveData provided by the Penthera SDK.
 */
public class MainActivity extends AppCompatActivity {

    // DEMO Server details
    private static final String BACKPLANE_PUBLIC_KEY = ;
    private static final String BACKPLANE_PRIVATE_KEY = ;

    // This is the first test asset which will be downloaded
    // Important: Asset ID should be unique across your video catalog
    private static final String ASSET_ID_1 = "TEST_ASSET_ID_1";
    private String ASSET_TITLE_1;
    private static final String ASSET_URL_1 = "https://storage.googleapis.com/wvmedia/clear/h264/tears/tears_sd.mpd";

    // This is the first test asset which will be downloaded
    // Important: Asset ID should be unique across your video catalog
    private static final String ASSET_ID_2 = "TEST_ASSET_ID_2";
    private String ASSET_TITLE_2;
    private static final String ASSET_URL_2 = "https://virtuoso-demo-content.s3.amazonaws.com/Steve/steve.m3u8";

    // This is the first test asset which will be downloaded
    // Important: Asset ID should be unique across your video catalog
    private static final String ASSET_ID_3 = "TEST_ASSET_ID_3";
    private String ASSET_TITLE_3;
    private static final String ASSET_URL_3 = "https://virtuoso-demo-content.s3.amazonaws.com/College/college.m3u8";

    // Important: The LiveData Factory object from the Penthera SDK enables the SDK to be lifecycle aware. It contains
    // a reference to the underlying Virtuoso object and provides all of the LiveData objects for the SDK.
    private VirtuosoLiveDataFactory virtuosoLiveDataFactory;

    // The Virtuoso object provides the top level interface to the Penthera SDK for all aspects apart from LiveData.
    private Virtuoso virtuoso;

    // The three download buttons
    private Button download1;
    private Button download2;
    private Button download3;

    private RecyclerView recyclerView;
    private AssetsAdapter adapter;

    // Callback interface used for binding to the list items
    private final AssetItemCallback assetCallback = new AssetItemCallback() {
        @Override
        public void onOpen(ISegmentedAsset asset) {
            Intent intent = new Intent(MainActivity.this, AssetDetailActivity.class)
                    .setAction(ACTION_VIEW)
                    .putExtra(AssetDetailActivity.ASSET_ID_KEY, asset.getAssetId());
            startActivity(intent);
        }

        @Override
        public void onDelete(ISegmentedAsset asset) {
            virtuoso.getAssetManager().delete(asset.getId());
        }

        @Override
        public void onPlay(ISegmentedAsset asset){
            // Unused from list view
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ASSET_TITLE_1 = getString(R.string.download1_name);
        ASSET_TITLE_2 = getString(R.string.download2_name);
        ASSET_TITLE_3 = getString(R.string.download3_name);

        // The live data factory must be created and then initialized with a lifecycle.
        virtuosoLiveDataFactory = VirtuosoLiveDataFactory.getInstance();
        virtuoso = virtuosoLiveDataFactory.createVirtuosoWithLifecycle(this, this);

        AssetListViewModel.Factory factory = new AssetListViewModel.Factory(getApplication(), virtuosoLiveDataFactory);
        AssetListViewModel model = new ViewModelProvider(this, factory).get(AssetListViewModel.class);

        model.getUiAssets().observe(this, assets -> {
            if (assets != null) {
                adapter.setAssetList(assets);
            }
        });

        // Set up the three download buttons
        download1 = findViewById(R.id.download_1);
        download2 = findViewById(R.id.download_2);
        download3 = findViewById(R.id.download_3);

        download1.setOnClickListener(v -> downloadAsset(0));
        download2.setOnClickListener(v -> downloadAsset(1));
        download3.setOnClickListener(v -> downloadAsset(2));

        // Set up the recyclerview
        recyclerView = findViewById(R.id.downloads_list);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        recyclerView.addItemDecoration(new DividerItemDecoration(this, RecyclerView.VERTICAL));
        adapter = new AssetsAdapter(assetCallback);
        recyclerView.setAdapter(adapter);
        // This prevents the button on the list items flickering due to the default animation
        recyclerView.getItemAnimator().setChangeDuration(0);
    }

    /**
     * Initialize the Penthera SDK and the ViewModel with SDK LiveData objects
     **/
    public void initVirtuosoSDK() {
        // This is current best practice for starting the SDK
        int status = virtuoso.getBackplane().getAuthenticationStatus();
        if (status != AuthenticationStatus.AUTHENTICATED) { // If not authenticated then execute sdk startup

            // Here we use the simplest login with hard coded values
            // This starts the SDK by registering it with the backplane. It is asynchronous. Success can be observed using the IEngineObserver interface.
            virtuoso.startup(
                    UUID.randomUUID().toString(),    // provide an appropriate unique user id. A random uuid is used here for demonstration purposes only
                    null,                                           // Optional additional device id to be associated with the user account. This is not the device id generated by the SDK
                    BACKPLANE_PUBLIC_KEY,                               // Substitute the real public backplane key here
                    BACKPLANE_PRIVATE_KEY,                              // Substitute the real private backplane key here
                    null                         // Push registration callback, this will be detailed in subsequent tutorials.
                    // This callback does not indicate that SDK startup is complete
            );
        }
    }

    /**
     * A simple method to download an asset from the three assets statically defined for the test app.
     * Demonstrates adding a basic MPG DASH or HLS asset.
     * @param index Index of asset to add to the download queue.
     */
    private void downloadAsset(int index) {
        initVirtuosoSDK();

        String url;
        String title;
        String assetId;
        switch (index) {
            case 0:
                url = ASSET_URL_1;
                title = ASSET_TITLE_1;
                assetId = ASSET_ID_1;
                break;
            case 1:
                url = ASSET_URL_2;
                title = ASSET_TITLE_2;
                assetId = ASSET_ID_2;
                break;
            case 2:
            default:
                url = ASSET_URL_3;
                title = ASSET_TITLE_3;
                assetId = ASSET_ID_3;
                break;
        }

        URL assetUrl;
        try {
            assetUrl = new URL(url);
        } catch (MalformedURLException mue) {
            Toast.makeText(this, "Problem with asset or ancillary URL", Toast.LENGTH_LONG).show();
            return;
        }

        // Simple switch between MPEG-DASH and HLS asset creation for demonstration purposes
        if (url.endsWith("mpd")) {
            // Creation of assets follows a builder pattern.
            // Create the parameters for the new asset.
            MPDAssetBuilder.MPDAssetParams params = new MPDAssetBuilder()
                    .assetId(assetId)          // REQUIRED PARAMETER unique asset ID of the new asset
                    .manifestUrl(assetUrl)
                    .assetObserver(new AssetParseObserver(this))
                    .addToQueue(true)
                    .desiredVideoBitrate(Integer.MAX_VALUE)
                    .withMetadata(title)
                    .build();

            // Add to the SDK
            virtuoso.getAssetManager().createMPDSegmentedAssetAsync(params);
        } else {
            HLSAssetBuilder.HLSAssetParams params = new HLSAssetBuilder()
                    .assetId(assetId)          // REQUIRED PARAMETER unique asset ID of the new asset
                    .manifestUrl(assetUrl)
                    .assetObserver(new AssetParseObserver(this))
                    .addToQueue(true)
                    .desiredVideoBitrate(Integer.MAX_VALUE)
                    .withMetadata(title)
                    .build();

            // Add to the SDK
            virtuoso.getAssetManager().createHLSSegmentedAssetAsync(params);
        }
    }

    /**
     * This class observes when an asset parse is complete.
     */
    static class AssetParseObserver implements ISegmentedAssetFromParserObserver {

        private AppCompatActivity activity;

        public AssetParseObserver(AppCompatActivity activity) {
            this.activity = activity;
        }

        @Override
        public void complete(final ISegmentedAsset asset, int error, boolean addedToQueue) {
            activity.runOnUiThread(() -> {

                if(asset != null && error == 0) {
                    Toast.makeText(activity, "Asset parsed and " + (addedToQueue ? "added" : "not added") + " to download queue", Toast.LENGTH_LONG  ).show();
                }
                else{
                    Toast.makeText(activity, "Error " + error + " while parsing asset", Toast.LENGTH_LONG).show();
                }
            });
        }
    }
}

File: CnCDemo/Tutorials/Kotlin/Download2Go7/src/main/java/com/penthera/download2go7/MainActivity.kt
package com.penthera.download2go7

import android.annotation.SuppressLint
import android.app.ActivityManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.penthera.common.Common.AssetStatus
import com.penthera.common.Common.AuthenticationStatus
import com.penthera.download2go7.databinding.ActivityMainBinding
import com.penthera.virtuososdk.Common
import com.penthera.virtuososdk.client.*
import com.penthera.virtuososdk.client.IService.IConnectionObserver
import com.penthera.virtuososdk.client.Observers.IEngineObserver
import com.penthera.virtuososdk.client.autodownload.PlaylistConfigBuilder
import com.penthera.virtuososdk.client.builders.HLSAssetBuilder
import java.net.URL
import java.util.*

class MainActivity : AppCompatActivity() {

    private lateinit var binding : ActivityMainBinding
    private lateinit var  virtuoso : Virtuoso
    var asset : IAsset? =  null
    private lateinit var queueObserver: AssetQueueObserver


    private var internalUpdate : Boolean = false
    //local reference to the Download2Go service
    private var download2GoService : IService? = null

    private var pauseRequested : Boolean = false
    private var resumeRequested : Boolean = false

    /**
     * Connection observer monitors when the service is bound
     */
     private var serviceConnectionObserver = object : IConnectionObserver {
        override fun connected() {
            download2GoService?.let{
                if(it.isBound){
                    try {
                        val status = it.status
                        internalUpdate = true
                        binding.pauseEngine.isChecked = status == Common.EngineStatus.PAUSED
                        internalUpdate = false
                    } catch (se: ServiceException) {
                        Log.d(
                            MainActivity::class.java.name,
                            "Service Exception on getting service status"
                        )
                    }
                }
            }
        }

        override fun disconnected() {}
    }

    private val enginePauseObserver: IEngineObserver = object : EngineObserver() {
        override fun engineStatusChanged(status: Int) {
            if (pauseRequested) {
                pauseRequested = !(status == Common.EngineStatus.PAUSED)
                updateUI()
            }
            if (resumeRequested) {
                resumeRequested = !(status != Common.EngineStatus.PAUSED)
                updateUI()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)


        initVirtuosoSDK(savedInstanceState)

        binding.download.setOnClickListener { downloadAsset() }
        binding.play.setOnClickListener { playAsset()}
        binding.delete.setOnClickListener { deleteAsset() }

        binding.showPlaylist.setOnClickListener { showPlaylist() }

        binding.pauseAsset.setOnCheckedChangeListener { _, isChecked ->  pauseAsset(isChecked) }
        binding.pauseEngine.setOnCheckedChangeListener { _, isChecked -> pauseEngine(isChecked) }

        updateUI()
    }

    private fun pauseAsset(pause : Boolean){
        asset?.let{
            if(pause){
                virtuoso.assetManager.pauseDownload(it)
            }
            else{
                virtuoso.assetManager.resumeDownload(it)
            }
        }
    }

    private fun pauseEngine(pause: Boolean){
        download2GoService?.let{
            if(it.isBound){
                if(pause && it.status != Common.EngineStatus.PAUSED){
                    pauseRequested = true
                    it.pauseDownloads()
                }
                else if(!pause && it.status == Common.EngineStatus.PAUSED){
                    resumeRequested = true
                    it.resumeDownloads()
                }
            }
        }
    }

    private fun initVirtuosoSDK(savedInstanceState: Bundle?) {

        virtuoso = Virtuoso(this)
        queueObserver = AssetQueueObserver(this)

        download2GoService = virtuoso.service

        //this is the current best practice for initializing the SDK
        if(savedInstanceState == null){//initial start of activity will have null saved instance state
            val status = virtuoso.backplane?.authenticationStatus
            if(status == AuthenticationStatus.NOT_AUTHENTICATED){//if not authenticated execute sdk startup
                //here we use the simplest login with hard coded values

                virtuoso.startup(
                    URL(BACKPLANE_URL),//substitute the proper backplane url for you implementation
                    // substitute the proper backplane url for your implementation
                    UUID.randomUUID()
                        .toString(),    // provide an appropriate unique user id. A random uuid is used here for demonstration purposes only
                    null, //Optional additional device id to be associated with the user account.  This is not the device id generated by the virtuoso SDK
                    BACKPLANE_PUBLIC_KEY,//Penthera demo public key.  Substitute the correct one.
                    BACKPLANE_PRIVATE_KEY,
                    object : IPushRegistrationObserver {
                        override fun onServiceAvailabilityResponse(
                            pushService: Int,
                            errorCode: Int
                        ) {
                            //callback for push registration.  this will be detailed in subsequent tutorials
                        }
                    }
                )
            }
        }

        //load asset if it has already been downloaded
        asset = getCurrentAsset()

    }

    private fun getCurrentAsset() : IAsset? {

        for(id in ASSET_LIST){
            val list : MutableList<IIdentifier>? = virtuoso.assetManager.getByAssetId(id)

            list?.let{
                if (it.isNotEmpty())
                    return  list[0] as IAsset
            }
        }

        return null
    }

    override fun onResume() {
        super.onResume()

        //resume the VirtuosoSDK on activity resume
        if(Build.VERSION.SDK_INT == Build.VERSION_CODES.P){
            //this is the recommended workaround for issuetracker.google.com/issues/110237673
            val activityManager = getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            val runningAppProcessInfo = activityManager.runningAppProcesses?.first { it.pid == android.os.Process.myPid() }
            if (runningAppProcessInfo != null && runningAppProcessInfo.importance >= ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND) {
                virtuoso.onResume()
            }
        }
        else{
            virtuoso.onResume()
        }
        virtuoso.addObserver(queueObserver)
        virtuoso.addObserver(enginePauseObserver)
        download2GoService?.setConnectionObserver(serviceConnectionObserver)
        download2GoService?.bind()


    }

    override fun onPause() {
        super.onPause()

        //pause the VirtuosoSDK on activity pause
        virtuoso.onPause()
        virtuoso.removeObserver(queueObserver)
        virtuoso.removeObserver(enginePauseObserver)
        download2GoService?.unbind()
        download2GoService?.setConnectionObserver(null)
    }

    private fun playAsset() {

        if(asset != null) {
            VideoPlayerActivity.playVideoDownload(asset!!, this)
        }


    }

    private fun deleteAsset() {
        virtuoso.assetManager.delete(asset)
        asset = null

        findViewById<TextView>(R.id.textView).text = "Deleting asset"
        val pb = findViewById<ProgressBar>(R.id.progressBar)
        pb.visibility = View.INVISIBLE
        pb.progress = 0

        updateUI()
    }

    private fun showPlaylist(){
        
        startActivity(Intent(this, PlaylistItemsActivity::class.java).apply {
            action = Intent.ACTION_VIEW
            putExtra(PlaylistItemsActivity.PLAYLIST_NAME, TEST_PLAYLIST_NAME)
        })

    }

    fun updateUI() {

        binding.download.isEnabled = asset == null
        binding.play.isEnabled = asset != null
        binding.delete.isEnabled = asset != null

        if( asset == null){
            binding.textView.text = ""
            binding.currentAsset.text = ""
            binding.pauseAsset.isEnabled = false
        }
        else{
            binding.pauseAsset.isEnabled = true
            binding.currentAsset.text = asset!!.metadata
        }
    }

    private fun downloadAsset(){

        virtuoso.assetManager.playlistManager.find(TEST_PLAYLIST_NAME)?.let{
            //clicking the download button will restart the entire autodownload process
            // so we clear the assets and history of the exisitng playlist
            virtuoso.assetManager.playlistManager.clear(TEST_PLAYLIST_NAME)
            //then add the assets we want in the playlist
            it.append(ASSET_LIST.toMutableList())
        } ?: run {
            //create playlist configuration options
            val plConfig = PlaylistConfigBuilder().apply{
                withName(TEST_PLAYLIST_NAME)
                requirePlayback(true)
                considerAssetHistory(false)
                searchFromBeginning(false)
            }.build()

            //create playlist with list of asset ids
            virtuoso.assetManager.playlistManager.create(plConfig, ASSET_LIST.toMutableList())
        }

        //create first asset in playlist
        val params = HLSAssetBuilder().apply {
            assetId(ASSET_ID)//REQUIRED PARAMETER asset ID of the new asset
            manifestUrl(URL(ASSET_URL))//REQUIRED PARAMETER  url of the new asset
            assetObserver(AssetParseObserver(this@MainActivity))//REQUIRED PARAMETER observer
            addToQueue(true)// add to the download queue after parsing complete
            desiredVideoBitrate(Int.MAX_VALUE)//specify a bitrate for desired video quality Integer.MAX_VALUE for largest available
            withMetadata(ASSET_TITLE)//user specified descriptive text for the asset.  Here we supply a title.
        }.build()

        virtuoso.assetManager.createHLSSegmentedAssetAsync(params)
    }

    class AssetParseObserver(activity : AppCompatActivity) : ISegmentedAssetFromParserObserver{

        private var mActivty : AppCompatActivity = activity

        @SuppressLint("ShowToast")
        override fun complete(asset: ISegmentedAsset?, error : Int, addedToQueue : Boolean) {

            if(asset != null){
               Toast.makeText(mActivty, "Asset parsed and " + if(addedToQueue) "added" else "not added" + "to download queue", Toast.LENGTH_LONG  ).show()

            }
            else{
                Toast.makeText(mActivty, "Error $error while parsing asset", Toast.LENGTH_LONG).show()
            }
        }
    }

    class AssetQueueObserver(activity: MainActivity) : Observers.IQueueObserver {


        private var lastProgress : Int = -1
        private var  mActivity : MainActivity = activity



        override fun engineStartedDownloadingAsset(aAsset: IIdentifier) {
            lastProgress = -1
            updateItem(aAsset, true)
        }

        override fun enginePerformedProgressUpdateDuringDownload(aAsset: IIdentifier) {
            updateItem(aAsset, true)
        }

        override fun engineCompletedDownloadingAsset(aAsset: IIdentifier) {
            updateItem(aAsset, true)
        }

        override fun engineEncounteredErrorDownloadingAsset(aAsset: IIdentifier) {
            // The base implementation does nothing.  See class documentation.
        }

        override fun engineUpdatedQueue() {
            // This indicates a change to the download queue - meaning either we added or removed something
            val assetManager: IAssetManager = mActivity.virtuoso.assetManager
            val queued = assetManager.queue.size()
            val downloaded = assetManager.downloaded.cursor.count
            val curAsset = mActivity.asset
            if ( curAsset != null && (queued > 0 || downloaded > 0)) {
                val asset = assetManager.get(curAsset.id) as IAsset
                if (asset.downloadStatus != curAsset.downloadStatus) {
                    mActivity.asset = asset
                    updateItem(asset, true)
                }
            }
            if (queued == 0) {
                // The asset has been deleted or downloaded
                mActivity.runOnUiThread {
                    mActivity.binding.textView.text = if (downloaded == 0) "Asset Deleted" else "Asset Downloaded"
                    mActivity.binding.progressBar.visibility = View.GONE
                }
            }
        }

        override fun engineEncounteredErrorParsingAsset(mAssetId: String) {}


        private fun updateItem(aFile: IIdentifier?, forceUpdate: Boolean) {

            aFile?.let {
                if(aFile is IAsset) {
                    val assetId = aFile.assetId

                    // Progress is for catalog item
                    if (assetId.isNotEmpty() && ASSET_LIST.contains(assetId)) {
                        //update our asset status
                        mActivity.runOnUiThread { updateItemStatus(aFile, forceUpdate) }

                    }
                }
            }
        }

        private fun updateItemStatus(asset: IAsset, forceUpdate: Boolean) {


            if(ASSET_LIST.contains(asset.assetId)){
                mActivity.asset = asset

                var progress  = (asset.currentSize / asset.expectedSize * 100).toInt()
                // Not a repeated progress -- Keep context switches minimal due to frequency of messages, unless forced
                if (forceUpdate || progress != lastProgress) {
                    val assetStatus : String
                    val fds = asset.downloadStatus
                    val value: String
                    when (fds) {

                        AssetStatus.DOWNLOADING -> {
                            assetStatus = mActivity.getString(R.string.status_downloading)
                            value = "downloading"
                        }

                        AssetStatus.DOWNLOAD_COMPLETE -> {
                            assetStatus = mActivity.getString(R.string.status_downloaded)
                            value = "complete"
                        }

                        AssetStatus.EXPIRED -> {
                            assetStatus = mActivity.getString(R.string.status_expired)
                            value = "expired"
                        }

                        AssetStatus.DOWNLOAD_DENIED_ASSET -> {
                            assetStatus = "Queued"
                            value = "DENIED : MAD"
                        }

                        AssetStatus.DOWNLOAD_DENIED_ACCOUNT -> {
                            assetStatus = "Queued"
                            value = "DENIED : MDA"
                        }

                        AssetStatus.DOWNLOAD_DENIED_EXTERNAL_POLICY -> {
                            assetStatus = "Queued"
                            value = "DENIED : EXT"
                        }

                        AssetStatus.DOWNLOAD_DENIED_MAX_DEVICE_DOWNLOADS -> {
                            assetStatus = "Queued"
                            value = "DENIED :MPD"
                        }

                        AssetStatus.DOWNLOAD_BLOCKED_AWAITING_PERMISSION -> {
                            assetStatus = "Queued"
                            value = "AWAITING PERMISSION"
                        }

                        AssetStatus.DOWNLOAD_DENIED_COPIES -> {
                            assetStatus = "Queued"
                            value = "DENIED : COPIES"
                        }

                        else -> {
                            assetStatus = mActivity.getString(R.string.status_pending)
                            value = "pending"
                        }
                    }


                    mActivity.updateUI()
                    val tv = mActivity.binding.textView
                    tv.visibility = View.VISIBLE
                    tv.text = String.format(mActivity.getString(R.string.asset_status), assetStatus, asset.errorCount, value)

                    lastProgress = progress
                    // Tiny Progress
                    if (progress == 0) progress = 1

                    // Progress Bar
                    val pb = mActivity.binding.progressBar
                    if (progress in 1..99) {
                        pb.progress = progress
                        pb.visibility = View.VISIBLE
                    } else {
                        pb.visibility = View.GONE
                    }
                }
            }

        }
    }

    companion object{
        // Important: Asset ID should be unique across your video catalog
        const val ASSET_ID : String = "SEASON-1-EPISODE-1"
        const val ASSET_TITLE : String = "Season 1 Episode 1"
        const val ASSET_ID2 : String = "SEASON-1-EPISODE-2"
        const val ASSET_TITLE2 : String = "Season 1 Episode 2"
        const val ASSET_ID3 : String = "SEASON-1-EPISODE-3"
        const val ASSET_TITLE3 : String = "Season 1 Episode 3"
        const val ASSET_ID4 : String = "SEASON-1-EPISODE-4"
        const val ASSET_TITLE4 : String = "Season 1 Episode 4"

        const val TEST_PLAYLIST_NAME : String = "TEST_PLAYLIST"

        const val ASSET_URL: String = "http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep1/index.m3u8"
        const val ASSET_URL2: String = "http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep2/index.m3u8"
        const val ASSET_URL3: String = "http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep3/index.m3u8"
        const val ASSET_URL4: String = "http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep4/index.m3u8"

        val ASSET_MAP = mapOf(ASSET_ID to Pair(ASSET_TITLE,  ASSET_URL),
            ASSET_ID2 to Pair(ASSET_TITLE2, ASSET_URL2),
            ASSET_ID3 to Pair(ASSET_TITLE3, ASSET_URL3),
            ASSET_ID4 to Pair(ASSET_TITLE4, ASSET_URL4))

        val ASSET_LIST = listOf(ASSET_ID, ASSET_ID2, ASSET_ID3, ASSET_ID4)

        const val BACKPLANE_URL = "https://demo.penthera.com"
        const val BACKPLANE_PUBLIC_KEY =  
        const val BACKPLANE_PRIVATE_KEY = 

    }
}
